module/VuFind/src/VuFind/Recommend/Factory.php

```diff
@@ -189,15 +189,18 @@ public static function getFavoriteFacets(ServiceManager $sm)
      */
     public function getMapSelection(ServiceManager $sm)
     {
-        $config = $sm->get('VuFind\Config\PluginManager');
         $backend = $sm->get('VuFind\Search\BackendManager');
         $solr = $backend->get('Solr');
 
         // add basemap options
         $basemapConfig = $sm->get('VuFind\GeoFeatures\BasemapConfig');
         $basemapOptions = $basemapConfig->getBasemap('MapSelection');
 
-        return new MapSelection($config, $solr, $basemapOptions);
+        // get MapSelection options
+        $mapSelectionConfig = $sm->get('VuFind\GeoFeatures\MapSelectionConfig');
```

Comment:
When I first tested this branch, I got a fatal error because there was a `$sm->getServiceLocator()->get()` call here, which no longer works following the upgrade to ServiceManager v3. I've gone ahead and fixed it, but just wanted to let you know. In the old SMv2, factories were passed a plugin-manager, and you had to use getServiceLocator to move up to the parent top-level service manager. In SMv3, the top-level service manager is always the one passed to factories, which makes things a lot more consistent and readable.

Reply:
Ok. Good to know and thanks for the catch and explanation.

Reply:
Fixing this didn't help with the bad request -- it resolved a 500 error that I was getting before I ran into the 400 error. I haven't tested the code since your latest revisions, though; I'm in meetings all day today, and tomorrow we're expecting a blizzard, but I'll try to give it another shot later in the week!

Reply:
Ok. Let me see if pulling out all the OpenLayers code and refactoring things fixes this. Let me see if I can reproduce it.

Reply:
I've mentioned that this problem was fixed elsewhere, but I'm also commenting on it here so we can see that this conversation is resolved when reviewing the PR in future. This is all better now! :-)